tenseal/tensealcontext.h

```diff
@@ -37,124 +37,38 @@ class TenSEALContext {
     static shared_ptr<TenSEALContext> Create(scheme_type scheme,
                                              size_t poly_modulus_degree,
                                              uint64_t plain_modulus,
-                                             vector<int> coeff_mod_bit_sizes) {
-        EncryptionParameters parms;
-        switch (scheme) {
-            case scheme_type::BFV:
-                parms = create_bfv_parameters(
-                    poly_modulus_degree, plain_modulus, coeff_mod_bit_sizes);
-                break;
+                                             vector<int> coeff_mod_bit_sizes);
+    static shared_ptr<TenSEALContext> Create(const char* filename);
 
-            case scheme_type::CKKS:
-                parms = create_ckks_parameters(poly_modulus_degree,
-                                               coeff_mod_bit_sizes);
-                break;
-
-            default:
-                throw invalid_argument("invalid scheme_type");
-        }
-
-        return shared_ptr<TenSEALContext>(new TenSEALContext(parms));
-    }
-
-    static shared_ptr<TenSEALContext> Create(const char* filename) {
-        return shared_ptr<TenSEALContext>(new TenSEALContext(filename));
-    }
-
-    PublicKey public_key() { return *this->_public_key; }
-    SecretKey secret_key() {
-        if (this->_secret_key == nullptr) {
-            throw invalid_argument(
-                "the current context is public, it doesn't hold a Secret key");
-        }
-
-        return *this->_secret_key;
-    }
-
-    RelinKeys relin_keys() {
-        if (this->_relin_keys == nullptr) {
-            throw invalid_argument(
-                "the current context doesn't hold a Relinearization keys");
-        }
-
-        return *this->_relin_keys;
-    }
-
-    GaloisKeys galois_keys() {
-        if (this->_galois_keys == nullptr) {
-            throw invalid_argument(
-                "the current context doesn't hold a Galois keys");
-        }
-
-        return *this->_galois_keys;
-    }
+    /*
+    Returns a pointer to the public key
+    */
+    shared_ptr<PublicKey> public_key();
+    shared_ptr<SecretKey> secret_key();
+    shared_ptr<RelinKeys> relin_keys();
+    shared_ptr<GaloisKeys> galois_keys();
```

Comment:
@philomath213 might argue why we shouldn't return shared_ptr for keys, what was the reason Bilal?

Reply:
I think there were some places with value copies on the galois keys, those can get huge, we shouldn't copy them.